src/UBAFeeCalculator/UBAFeeSpokeCalculatorAnalog.ts

```diff
@@ -0,0 +1,202 @@
+/** Provides the static, procedural versions of the functions provided in the UBAFeeSpokeCalculator */
+
+import { BigNumber } from "ethers";
+import { UbaFlow, isUbaInflow } from "../interfaces";
+import { TokenRunningBalanceWithNetSend, UBAFlowFee } from "./UBAFeeTypes";
+import UBAConfig from "./UBAFeeConfig";
+import { toBN } from "../utils";
+import { computePiecewiseLinearFunction } from "./UBAFeeUtility";
+
+/**
+ * Calculates the running balance for a given token on a spoke chain
+ * @param flows The flows to calculate the running balance for
+ * @param lastValidatedRunningBalance The last validated running balance for the token
+ * @param lastValidatedIncentiveRunningBalance The last validated incentive running balance for the token
+ * @param chainId The chain id of the spoke chain
+ * @param tokenSymbol The token symbol of the token to calculate the running balance for
+ * @param config The UBAConfig to use for the calculation
+ * @returns The running balance for the token
+ */
+export function calculateHistoricalRunningBalance(
+  flows: UbaFlow[],
+  lastValidatedRunningBalance: BigNumber,
+  lastValidatedIncentiveRunningBalance: BigNumber,
+  chainId: number,
+  tokenSymbol: string,
+  config: UBAConfig
+): TokenRunningBalanceWithNetSend {
+  // Attempt to resolve the trigger hurdle to include in the running
+  // balance calculation
+  const { upperBound: upperBoundTriggerHurdle, lowerBound: lowerBoundTriggerHurdle } =
+    config.getBalanceTriggerThreshold(chainId, tokenSymbol);
+
+  // If the last validated running balance is undefined, we need to compute the running balance from scratch
+  // This is the case when the UBA Fee Calculator is first initialized or run on a range
+  // that we haven't computed the running balance for yet
+  const historicalResult: TokenRunningBalanceWithNetSend = flows.reduce(
+    (acc, flow) => {
+      // If the flow is an inflow, we need to add the amount to the running balance
+      // If the flow is an outflow, we need to subtract the amount from the running balance
+      // This is reflected in the incentive balance as well
+      const resultant: TokenRunningBalanceWithNetSend = {
+        netRunningBalanceAdjustment: toBN(acc.netRunningBalanceAdjustment.toString()), // Deep copy via string conversion
+        runningBalance: acc.runningBalance[isUbaInflow(flow) ? "add" : "sub"](flow.amount),
```

Comment:
Where should we compute the incentive fee and subtract it from this line and add it in next line?

Reply:
We are now accounting for the incentive fee.

Commit: https://github.com/across-protocol/sdk-v2/pull/266/commits/340e5d554ccdd39c0b0a4abd29c0466f0e36c864